File: samples/BlobExample.java
package com.example.samples;

import java.util.List;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.Map;
import java.sql.Connection;
import java.sql.Statement;
import java.sql.ResultSet;

/**
 * OrderManager - Example of Blob (God Class) code smell
 * This class has too many responsibilities and should be split
 */
public class OrderManager {
    private Connection database;
    private EmailService emailService;
    private Logger logger;
    private PaymentGateway paymentGateway;
    private InventorySystem inventory;
    private Map<String, Order> orderCache;
    
    public OrderManager() {
        this.orderCache = new HashMap<>();
        this.logger = new Logger("OrderManager");
    }
    
    // Order CRUD operations
    public void createOrder(Order order) {
        try {
            validateOrder(order);
            saveToDatabase(order);
            updateInventory(order);
            sendOrderConfirmation(order);
            orderCache.put(order.getId(), order);
            logger.log("Order created: " + order.getId());
        } catch (Exception e) {
            logger.error("Failed to create order", e);
        }
    }
    
    public void updateOrder(Order order) {
        if (!orderCache.containsKey(order.getId())) {
            logger.warn("Order not in cache: " + order.getId());
        }
        try {
            Statement stmt = database.createStatement();
            stmt.execute("UPDATE orders SET status='" + order.getStatus() + 
                        "' WHERE id='" + order.getId() + "'");
            orderCache.put(order.getId(), order);
        } catch (Exception e) {
            logger.error("Update failed", e);
        }
    }
    
    public void deleteOrder(String orderId) {
        try {
            Statement stmt = database.createStatement();
            stmt.execute("DELETE FROM orders WHERE id='" + orderId + "'");
            orderCache.remove(orderId);
            logger.log("Order deleted: " + orderId);
        } catch (Exception e) {
            logger.error("Delete failed", e);
        }
    }
    
    public List<Order> findOrders(String criteria) {
        List<Order> results = new ArrayList<>();
        try {
            Statement stmt = database.createStatement();
            ResultSet rs = stmt.executeQuery("SELECT * FROM orders WHERE " + criteria);
            while (rs.next()) {
                Order order = new Order();
                order.setId(rs.getString("id"));
                order.setCustomerId(rs.getString("customer_id"));
                order.setStatus(rs.getString("status"));
                results.add(order);
            }
        } catch (Exception e) {
            logger.error("Find failed", e);
        }
        return results;
    }
    
    // Payment processing
    public boolean processPayment(Payment payment) {
        try {
            if (payment.getAmount() <= 0) {
                return false;
            }
            boolean result = paymentGateway.charge(payment);
            if (result) {
                updateOrderStatus(payment.getOrderId(), "PAID");
                sendPaymentConfirmation(payment);
            }
            return result;
        } catch (Exception e) {
            logger.error("Payment processing failed", e);
            return false;
        }
    }
    
    public void refundPayment(String paymentId) {
        try {
            Payment payment = findPayment(paymentId);
            paymentGateway.refund(payment);
            updateOrderStatus(payment.getOrderId(), "REFUNDED");
            sendRefundNotification(payment);
        } catch (Exception e) {
            logger.error("Refund failed", e);
        }
    }
    
    // Notification handling
    public void sendOrderConfirmation(Order order) {
        String message = "Your order " + order.getId() + " has been confirmed.";
        emailService.send(order.getCustomerEmail(), "Order Confirmation", message);
    }
    
    public void sendShippingNotification(Order order) {
        String message = "Your order " + order.getId() + " has been shipped.";
        emailService.send(order.getCustomerEmail(), "Shipping Update", message);
    }
    
    public void sendPaymentConfirmation(Payment payment) {
        String message = "Payment received: $" + payment.getAmount();
        emailService.send(payment.getCustomerEmail(), "Payment Confirmation", message);
    }
    
    public void sendRefundNotification(Payment payment) {
        String message = "Refund processed: $" + payment.getAmount();
        emailService.send(payment.getCustomerEmail(), "Refund Processed", message);
    }
    
    // Reporting
    public Report generateSalesReport(String startDate, String endDate) {
        Report report = new Report();
        try {
            Statement stmt = database.createStatement();
            ResultSet rs = stmt.executeQuery(
                "SELECT SUM(amount) as total FROM orders WHERE date BETWEEN '" + 
                startDate + "' AND '" + endDate + "'"
            );
            if (rs.next()) {
                report.setTotalSales(rs.getDouble("total"));
            }
            report.setTitle("Sales Report: " + startDate + " to " + endDate);
        } catch (Exception e) {
            logger.error("Report generation failed", e);
        }
        return report;
    }
    
    public Report generateInventoryReport() {
        Report report = new Report();
        List<InventoryItem> items = inventory.getAllItems();
        for (InventoryItem item : items) {
            report.addLine(item.getName() + ": " + item.getQuantity());
        }
        return report;
    }
    
    // Validation
    public boolean validateOrder(Order order) {
        if (order.getItems() == null || order.getItems().isEmpty()) {
            logger.error("Order has no items");
            return false;
        }
        for (OrderItem item : order.getItems()) {
            if (item.getQuantity() <= 0) {
                logger.error("Invalid quantity for item: " + item.getProductId());
                return false;
            }
        }
        return true;
    }
    
    public boolean validatePayment(Payment payment) {
        if (payment.getAmount() <= 0) {
            return false;
        }
        if (payment.getCardNumber() == null || payment.getCardNumber().length() < 13) {
            return false;
        }
        return true;
    }
    
    // Database operations
    public void connectDatabase(String url, String username, String password) {
        try {
            database = java.sql.DriverManager.getConnection(url, username, password);
            logger.log("Database connected");
        } catch (Exception e) {
            logger.error("Database connection failed", e);
        }
    }
    
    public void closeDatabase() {
        try {
            if (database != null) {
                database.close();
            }
        } catch (Exception e) {
            logger.error("Database close failed", e);
        }
    }
    
    // Helper methods
    private void saveToDatabase(Order order) throws Exception {
        // Implementation
    }
    
    private void updateInventory(Order order) {
        // Implementation
    }
    
    private void updateOrderStatus(String orderId, String status) {
        // Implementation
    }
    
    private Payment findPayment(String paymentId) {
        // Implementation
        return null;
    }
}

// Supporting classes
class Order {
    private String id;
    private String customerId;
    private String customerEmail;
    private String status;
    private List<OrderItem> items;
    
    public String getId() { return id; }
    public void setId(String id) { this.id = id; }
    public String getCustomerId() { return customerId; }
    public void setCustomerId(String customerId) { this.customerId = customerId; }
    public String getCustomerEmail() { return customerEmail; }
    public String getStatus() { return status; }
    public void setStatus(String status) { this.status = status; }
    public List<OrderItem> getItems() { return items; }
}

class OrderItem {
    private String productId;
    private int quantity;
    
    public String getProductId() { return productId; }
    public int getQuantity() { return quantity; }
}

class Payment {
    private String id;
    private String orderId;
    private String customerId;
    private String customerEmail;
    private String cardNumber;
    private double amount;
    
    public String getId() { return id; }
    public String getOrderId() { return orderId; }
    public String getCustomerId() { return customerId; }
    public String getCustomerEmail() { return customerEmail; }
    public String getCardNumber() { return cardNumber; }
    public double getAmount() { return amount; }
}

class Report {
    private String title;
    private double totalSales;
    private List<String> lines;
    
    public Report() {
        this.lines = new ArrayList<>();
    }
    
    public void setTitle(String title) { this.title = title; }
    public void setTotalSales(double total) { this.totalSales = total; }
    public void addLine(String line) { this.lines.add(line); }
}

class EmailService {
    public void send(String to, String subject, String message) {}
}

class Logger {
    public Logger(String name) {}
    public void log(String message) {}
    public void error(String message) {}
    public void error(String message, Exception e) {}
    public void warn(String message) {}
}

class PaymentGateway {
    public boolean charge(Payment payment) { return true; }
    public void refund(Payment payment) {}
}

class InventorySystem {
    public List<InventoryItem> getAllItems() { return new ArrayList<>(); }
}

class InventoryItem {
    private String name;
    private int quantity;
    
    public String getName() { return name; }
    public int getQuantity() { return quantity; }
}

File: samples js/BlobExample.js
/** OrderManager — handles orders, payments, notifications, reporting, and storage */
class OrderManager {
    constructor() {
        this.orderCache = new Map();
        this.database = null;
        this.emailService = null;
        this.paymentGateway = null;
        this.inventory = null;
    }

    // Order CRUD operations
    createOrder(order) {
        try {
            this.validateOrder(order);
            this.saveToDatabase(order);
            this.updateInventory(order);
            this.sendOrderConfirmation(order);
            this.orderCache.set(order.id, order);
            console.log("Order created: " + order.id);
        } catch (e) {
            console.error("Failed to create order", e);
        }
    }

    updateOrder(order) {
        if (!this.orderCache.has(order.id)) {
            console.warn("Order not in cache: " + order.id);
        }
        try {
            this.database.query(
                "UPDATE orders SET status='" + order.status + "' WHERE id='" + order.id + "'"
            );
            this.orderCache.set(order.id, order);
        } catch (e) {
            console.error("Update failed", e);
        }
    }

    deleteOrder(orderId) {
        try {
            this.database.query("DELETE FROM orders WHERE id='" + orderId + "'");
            this.orderCache.delete(orderId);
            console.log("Order deleted: " + orderId);
        } catch (e) {
            console.error("Delete failed", e);
        }
    }

    findOrders(criteria) {
        const results = [];
        try {
            const rows = this.database.query("SELECT * FROM orders WHERE " + criteria);
            for (const row of rows) {
                results.push({
                    id: row.id,
                    customerId: row.customer_id,
                    status: row.status,
                });
            }
        } catch (e) {
            console.error("Find failed", e);
        }
        return results;
    }

    // Payment processing
    processPayment(payment) {
        try {
            if (payment.amount <= 0) return false;
            const result = this.paymentGateway.charge(payment);
            if (result) {
                this.updateOrderStatus(payment.orderId, "PAID");
                this.sendPaymentConfirmation(payment);
            }
            return result;
        } catch (e) {
            console.error("Payment processing failed", e);
            return false;
        }
    }

    refundPayment(paymentId) {
        try {
            const payment = this.findPayment(paymentId);
            this.paymentGateway.refund(payment);
            this.updateOrderStatus(payment.orderId, "REFUNDED");
            this.sendRefundNotification(payment);
        } catch (e) {
            console.error("Refund failed", e);
        }
    }

    // Notification handling
    sendOrderConfirmation(order) {
        const message = "Your order " + order.id + " has been confirmed.";
        this.emailService.send(order.customerEmail, "Order Confirmation", message);
    }

    sendShippingNotification(order) {
        const message = "Your order " + order.id + " has been shipped.";
        this.emailService.send(order.customerEmail, "Shipping Update", message);
    }

    sendPaymentConfirmation(payment) {
        const message = "Payment received: $" + payment.amount;
        this.emailService.send(payment.customerEmail, "Payment Confirmation", message);
    }

    sendRefundNotification(payment) {
        const message = "Refund processed: $" + payment.amount;
        this.emailService.send(payment.customerEmail, "Refund Processed", message);
    }

    // Reporting
    generateSalesReport(startDate, endDate) {
        const report = { title: "", totalSales: 0, lines: [] };
        try {
            const rows = this.database.query(
                "SELECT SUM(amount) as total FROM orders WHERE date BETWEEN '" +
                startDate + "' AND '" + endDate + "'"
            );
            if (rows.length > 0) {
                report.totalSales = rows[0].total;
            }
            report.title = "Sales Report: " + startDate + " to " + endDate;
        } catch (e) {
            console.error("Report generation failed", e);
        }
        return report;
    }

    generateInventoryReport() {
        const report = { title: "Inventory Report", lines: [] };
        const items = this.inventory.getAllItems();
        for (const item of items) {
            report.lines.push(item.name + ": " + item.quantity);
        }
        return report;
    }

    // Validation
    validateOrder(order) {
        if (!order.items || order.items.length === 0) {
            console.error("Order has no items");
            return false;
        }
        for (const item of order.items) {
            if (item.quantity <= 0) {
                console.error("Invalid quantity for item: " + item.productId);
                return false;
            }
        }
        return true;
    }

    validatePayment(payment) {
        if (payment.amount <= 0) return false;
        if (!payment.cardNumber || payment.cardNumber.length < 13) return false;
        return true;
    }

    // Database operations
    connectDatabase(url, username, password) {
        try {
            this.database = new DatabaseClient(url, username, password);
            console.log("Database connected");
        } catch (e) {
            console.error("Database connection failed", e);
        }
    }

    closeDatabase() {
        try {
            if (this.database) this.database.close();
        } catch (e) {
            console.error("Database close failed", e);
        }
    }

    // Helper methods
    saveToDatabase(order) { /* implementation */ }
    updateInventory(order) { /* implementation */ }
    updateOrderStatus(orderId, status) { /* implementation */ }
    findPayment(paymentId) { return null; }
}

module.exports = { OrderManager };
